asllib/types.ml

```diff
@@ -346,173 +346,126 @@ module Domain = struct
 
   let compare _d1 _d2 = assert false
 
-  module Iterators = struct
+  (** The [StaticApprox] module creates constant approximation of integer
+      constraints as sets of integers. *)
+  module StaticApprox = struct
+    (** The two possible types of approximations. *)
     type approx = Over | Under
 
     exception CannotOverApproximate
-
-    let assert_under approx acc =
-      if approx = Over then raise CannotOverApproximate else acc
-
-    let fold_zs =
-      let rec aux f z1 z2 acc =
-        if Z.lt z2 z1 then acc else f z1 acc |> aux f (Z.succ z1) z2
+    (** Raised if over approximation is not possible. *)
+
+    (** Util to raise [CannotOverApproximate]. *)
+    let assert_under approx =
+      if approx = Over then raise CannotOverApproximate else IntSet.empty
+
+    let make_interval approx z1 z2 =
+      if Z.leq z1 z2 then IntSet.(add (Interval.make z1 z2) empty)
+      else assert_under approx
+
+    let intset_map f s =
+      IntSet.elements_individual s |> List.map f |> IntSet.of_list
+
+    let intset_cross_map f s1 s2 =
+      let l1 = IntSet.elements_individual s1
+      and l2 = IntSet.elements_individual s2 in
+      list_cross f l1 l2 |> IntSet.of_list
```

Comment:
Hi @HadrienRenaud, may I suggest the following half deforestested versions that would spare some intermediate lists:
```suggestion
    let intset_map f s =
      IntSet.fold_individual (fun x acc -> f x::acc) s [] |> IntSet.of_list

    let intset_cross_map f s1 s2 =
      IntSet.fold_individual
        (fun x acc ->
           IntSet.fold_individual
             (fun y acc -> f x y::acc)
             s2 acc)
        s1 [] |> IntSet.of_list
```